core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+use {
+    super::{
+        in_flight_tracker::InFlightTracker,
+        scheduler_error::SchedulerError,
+        thread_aware_account_locks::{ThreadAwareAccountLocks, ThreadId, ThreadSet},
+        transaction_priority_id::TransactionPriorityId,
+        transaction_state::SanitizedTransactionTTL,
+        transaction_state_container::TransactionStateContainer,
+    },
+    crate::banking_stage::{
+        consumer::TARGET_NUM_TRANSACTIONS_PER_BATCH,
+        multi_iterator_scanner::{
+            MultiIteratorScanner, PayloadAndAlreadyHandled, ProcessingDecision,
+        },
+        read_write_account_set::ReadWriteAccountSet,
+        scheduler_messages::{ConsumeWork, TransactionBatchId, TransactionId},
+    },
+    crossbeam_channel::Sender,
+    itertools::Itertools,
+    solana_sdk::{clock::Slot, transaction::SanitizedTransaction},
+};
+
+const QUEUED_TRANSACTION_LIMIT: usize = 64 * 100;
+
+/// Interface to perform scheduling for consuming transactions.
+/// Using a multi-iterator approach.
+pub struct MultiIteratorConsumeScheduler {
+    in_flight_tracker: InFlightTracker,
+    account_locks: ThreadAwareAccountLocks,
+    consume_work_senders: Vec<Sender<ConsumeWork>>,
+}
+
+impl MultiIteratorConsumeScheduler {
+    pub fn new(consume_work_senders: Vec<Sender<ConsumeWork>>) -> Self {
+        let num_threads = consume_work_senders.len();
+        Self {
+            in_flight_tracker: InFlightTracker::new(num_threads),
+            account_locks: ThreadAwareAccountLocks::new(num_threads),
+            consume_work_senders,
+        }
+    }
+
+    pub(crate) fn schedule(
+        &mut self,
+        container: &mut TransactionStateContainer,
+    ) -> Result<usize, SchedulerError> {
+        let num_threads = self.consume_work_senders.len();
+        let mut schedulable_threads = ThreadSet::any(num_threads);
+        let outstanding = self.in_flight_tracker.num_in_flight_per_thread();
+        for (thread_id, outstanding_count) in outstanding.iter().enumerate() {
+            if *outstanding_count > QUEUED_TRANSACTION_LIMIT {
+                schedulable_threads.remove(thread_id);
+            }
+        }
+
+        let active_scheduler = ActiveMultiIteratorConsumeScheduler {
+            in_flight_tracker: &mut self.in_flight_tracker,
+            account_locks: &mut self.account_locks,
+            batch_account_locks: ReadWriteAccountSet::default(),
+            batches: Batches::new(num_threads),
+            schedulable_threads,
+            unschedulable_transactions: Vec::new(),
+            batch_priorty_order_guard: ReadWriteAccountSet::default(),
+            pass_priority_order_guard: ReadWriteAccountSet::default(),
+        };
+
+        const MAX_TRANSACTIONS_PER_SCHEDULING_PASS: usize = 100_000;
+        let target_scanner_batch_size: usize = TARGET_NUM_TRANSACTIONS_PER_BATCH * num_threads;
+        let ids = container
+            .take_top_n(MAX_TRANSACTIONS_PER_SCHEDULING_PASS)
```

Comment:
The current banking_stage has 700_000 buffer limit, so theoretically that's the limit per pass. Wounding why 100_000 is chosen here?

Reply:
Would taking 10K results too much overhead? Otherwise, it'd be even faster, and likely visits priority queue multiple times per slot (to give prioritized txs more chances for landing).

Reply:
I can test w/ using 10k, since that's **probably** large enough.
We do want it to be a fairly large buffer because the MI will look ahead for non-conflicting txs.
So if we have some really competitive event (NFT mints for example), we may have nearly all our top-of-queue txs conflicting, MI would "look past" them and schedule some non-conflicting txs.

100k is nearly always going to have some non-conflicting txs - 10k will also probably nearly always have non-conflicting as well.